src/katsdpmodels/fetch/__init__.py

```diff
@@ -0,0 +1,260 @@
+            if url in self._alias_cache:
+                new_url = self._alias_cache[url]
+            else:
+                request = Request(url, ResponseType.TEXT)
+                response = yield request
+                assert isinstance(response, TextResponse)
+                rel_path = response.text.rstrip()
+                new_url = urllib.parse.urljoin(response.url, rel_path)
+                new_scheme = urllib.parse.urlparse(new_url).scheme
+                if new_scheme == 'file' and parts.scheme != 'file':
+                    raise models.LocalRedirectError.with_urls(
+                        f'Remote {url} tried to redirect to local {new_url}',
+                        url=url, original_url=chain[0])
+                self._alias_cache[url] = new_url
+            if new_url in chain:
+                raise models.TooManyAliasesError.with_urls(
+                    f'Cycle detected starting from {new_url}',
+                    url=new_url, original_url=chain[0])
+            chain.append(new_url)
+            _logger.debug('Redirecting from %s to %s', url, new_url)
+            url = new_url
+            parts = urllib.parse.urlparse(url)
+        return chain
+
+    def _handle_file_scheme(self, request: Request, lazy: bool = False) -> Response:
+        """Handle a request with ``file://`` scheme.
+
+        Subclasses can delegate to this function to deal with such URLs.
+        """
+        assert request.response_type in {ResponseType.TEXT, ResponseType.FILE}
+        parts = urllib.parse.urlparse(request.url)
+        path = pathlib.Path(urllib.parse.unquote(parts.path)).resolve()
+        response: Response
+        if request.response_type == ResponseType.TEXT:
+            with open(path, 'r', errors='replace') as f:
+                text = f.read()
+                response = TextResponse(path.as_uri(), {}, text)
+        else:
+            with contextlib.ExitStack() as exit_stack:
+                file = open(path, 'rb')
+                exit_stack.callback(file.close)
+                if not lazy and file.seekable:
+                    content: Optional[bytes] = file.read()
+                    file.seek(0)
+                else:
+                    content = None
+                exit_stack.pop_all()
+                response = FileResponse(path.as_uri(), {}, cast(io.IOBase, file), content)
+        return response
+
+    def _get(self, url: str, model_class: Type[_M]) -> Generator[Request, Response, _M]:
+        original_url = url
+        url = (yield from self._resolve(url))[-1]
+        if url in self._model_cache:
+            model = self._model_cache[url]
+            if model_class.model_type != model.model_type:
+                raise models.ModelTypeError.with_urls(
+                    f'Expected a model of type {model_class.model_type!r}, '
+                    f'not {model.model_type!r}',
+                    url=url, original_url=original_url)
+            if not isinstance(model, model_class):
```

Comment:
In this case it indicates programmer error, not just a model of the wrong type. Either the programmer has passed in a subclass instead of the base class (e.g. `RFIMaskRanges` instead of `RFIMask`) or somehow there are two different "base" classes for the model type.